geode-core/src/main/java/org/apache/geode/internal/cache/AbstractUpdateOperation.java

```diff
@@ -144,16 +144,21 @@ public static boolean doPutOrCreate(LocalRegion rgn, EntryEventImpl ev, long las
           br.getPartitionedRegion().getPrStats().startApplyReplication();
         }
         try {
-          // if the oldValue is the DESTROYED token and overwrite is disallowed,
-          // then basicPut will set the blockedDestroyed flag in the event
-          boolean overwriteDestroyed = ev.getOperation().isCreate();
-          if (rgn.basicUpdate(ev, true /* ifNew */, false/* ifOld */, lastMod,
-              overwriteDestroyed)) {
-            rgn.getCachePerfStats().endPut(startPut, ev.isOriginRemote());
-            // we did a create, or replayed a create event
-            doUpdate = false;
-            updated = true;
-          } else { // already exists. If it was blocked by the DESTROYED token, then
+          if (rgn.getRegionEntry(ev.getKey()) == null) {
```

Comment:
getRegionEntry returning null does mean we should try a create.
But getRegionEntry returning non-null does not mean we should skip the create and do an update.
The non-null RegionEntry can be one that contains a tombstone in which case we should still do a create.

I think it introduces race conditions to try to fix this here. The problem is that by the time you see the region entry and decide to do something based on that state, some other thread can change the state our from under you.

The previous code does not look thread safe to me either. But I think the way to fix GEODE-6755 is to enhance basicUpdate so that it can be called once to do either a create or update. basicUpdate has the code that syncs the region entry and can then atomically figure out if the current op is a create or update.